### PR TITLE
In Dev Cascade, rewrite OmniNav Off-Canvas Nav using faster index block.

### PR DESCRIPTION
**Pull request by tatwell**
Trello Card: https://trello.com/c/v5qK35x8/

I ended up not testing with the new index block. I assume new index block will not work with off-canvas nav.

I included new refactored OmniNav format and old one it will replace. I'd like this refactored format to serve as a reference for format coding style going forward. So please comment regarding anything you don't like or don't understant.

### CODE REVIEW

File: .cascade-code/Chapman.edu/_cascade/formats/Omni-Nav.vtl [hunk 1]
```diff
@@ -0,0 +1,639 @@
+#*
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_cascade/formats/Omni-Nav.vtl]: This is the existing Omni-Nav format.

New refactored one should behave just like this. I tested it on Dev (see Trello card comments for details) and AFAICT it does.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl [hunk 2]
```diff
@@ -0,0 +1,37 @@
+##
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl]: This is just a test format that can be used for page content. It demonstrates some basic debug helpers and Velocity collection datatypes.

Note `.vtl` extension. Going forward, I propose all Velocity formats in Cascade include this in their name and follow Rails snake_case file naming convention.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+  [ 'admission', 'Admission', 'default', '/' ],
+  [ 'alumni', 'Alumni', 'default', '/' ],
+  [ 'campus-services', 'Campus Services', 'default', '/' ],
+  [ 'faculty-staff', 'Faculty and Staff', 'default', '/' ],
+  [ 'families', 'Families', 'default', '/' ],
+  [ 'research', 'Research', 'default', '/' ],
+  [ 'students', 'Students', 'default', '/' ],
+  [ 'support-chapman', 'Support Chapman', 'default', '/' ],
+  [ 'communication', 'SCHOOL OF COMM', 'soc', 'communication/index.aspx' ]
+])
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: 4 lists from original one consolidate in a list of lists. Considered a map, but I wasn't sure order mattered. Anyway, this is probably simpler.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+          </ul>
+        #end
+      </li>
+    #end
+  #end
+#end
+
+
+##
+## Main OmniNav HTML Block
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: I like this "main block" pattern with most functionality encapsulated in macros.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+    #if ( $pagePath.contains($dirName) && $pagePathDir == $dirName )
+      #set ( $pageUmbrellaCategory = $umbrellaCategory )
+    #end
+  #end
+#end
+
+#macro ( loginLogoLink $href $svg_id $label )
+  ## Cascade won't allow this line to be saved when SVG_ID replaced by $svg_id.
+  ## Won't even let you save it with line commented out. Error message:
+  ## Draft could not be saved. named capturing group is missing trailing '}'
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: I'll remove this comment. Problem was hash was inside `system-asset` block.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl (hunk 2 above)
Comment (mirabelrice) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl]: 👍

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl (hunk 4 above)
Comment (mirabelrice) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: 👍

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+      ## Count hidden children to see if we need to show toggle button.
+      #set ( $hiddenChildCounter = 0 )
+      #set ( $hiddenChildSelector = "dynamic-metadata[1][value='Yes'] | system-folder/dynamic-metadata/value" )
+      #foreach ( $child in $children )
+        #if ( $_XPathTool.selectSingleNode($child, $hiddenChildSelector) )
+          #set ( $hiddenChildCounter = $hiddenChildCounter + 1 )
+        #end
+      #end
+      #set ( $hasVisibleChildren = $children.size() > 0 && $hiddenChildCounter < $children.size() )
+
```
Comment (kaylaziegler) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: This is what we use in the left nav format for hidden children:
```
#set ( $hiddenChildren = $_XPathTool.selectNodes($sibling, "system-page[name != 'index']/dynamic-metadata[1][value='Yes'] | system-folder[dynamic-metadata[1][value='Yes']]") )
        
#if ( $children.size() > $hiddenChildren.size() )
  #set ( $siblingHasChildren = true )
#else
  #set ( $siblingHasChildren = false )
#end
```
Would this not work here as well?

Also I do like `#set ( $hasVisibleChildren = $children.size() > 0 && $hiddenChildCounter < $children.size() )` and something like that would replace the if/else in the code I pasted above but there was a comment (from Mandy maybe?) that said it doesn't work to set a variable equal to a conditional. This is a comment from the old LeftNavNoSiblings format: `### DIDN't WORK: set ( $siblingHasChildren = $children.size() > 0 )`

Comment (tatwell) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: It did cross my mind that there might be issues with that conditional expression. It seemed to work in my tests. But I'll explicitly test and report back.

`selectNodes` code does look cleaner than the loop. I'll change this.

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl (hunk 2 above)
Comment (kaylaziegler) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl]: The left nav formats I wrote are not under `Chapman.edu/` in `.cascade-code`, should we have an action item or something to move them? They are under that folder in Cascade obviously, I just didn't copy that part to .cascade-code which I probably should have.

Comment (tatwell) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/debug_format.vtl]: Yes, add an action item for it. It's not urgent.

It's very Java-esque to include it but that's how Cascade lists out the path so let's include it and try to be consistent.

**Comment by kaylaziegler** [comment]
Added action item to move `.cascade-code/_components/` folder to `.cascade-code/Chapman.edu/_components` after this PR is merged. https://trello.com/c/2u36T5fK

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl (hunk 6 above)
Comment (mandythomas) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: remember that we hide all of the following: children/sibling pages marked as Hidden, and the index pages in the subfolders if marked as Hidden (remember they are children of the index page in parent folder), AND if a folder itself is marked as Hidden then no pages or subfolders/index-pages under it should show.

Comment (tatwell) [reply .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: @kaylaziegler It looks like conditional assignment does work. See latest changeset:

- https://github.com/chapmanu/cascade-assets/pull/243/commits/71a63c8d250e1778ca103039beacfe2d319a8eca#diff-c9f496e65abd2235fce97c7989dae13bR69

Results can be viewed here (bottom of page):

- https://dev-www.chapman.edu/test-section/tatwell/fast-omni-nav/demo.html.aspx

File: .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    #elseif ( $nodeType == 'system-folder' )
+      #set ( $folderIndexPage = $_XPathTool.selectSingleNode($siblingNode, "system-page[name = 'index']") )
+      #set ( $displayName = $_EscapeTool.xml($folderIndexPage.getChild("display-name").value) )
+      #set ( $children = $_XPathTool.selectNodes($siblingNode, "system-page[name != 'index'] | system-folder") )
+
+      ## If folder has visible children, we need to show toggle button with drop-down.
+      #set ( $hiddenChildSelector = "system-page[name != 'index']/dynamic-metadata[1][value='Yes'] | system-folder[dynamic-metadata[1][value='Yes']]" )
+      #set ( $hiddenChildren = $_XPathTool.selectNodes($siblingNode, $hiddenChildSelector) )
+      #set ( $hasVisibleChildren = $children.size() > 0 && $hiddenChildren.size() < $children.size() )
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/test-section/tatwell/fast-omni-nav/omni_nav.vtl]: Revised to user selectNodes rather than loop on children.

**Comment by tatwell** [comment]
Final call for comments on this PR. I'd like to merge and wrap up this story.

**Pull request merged**